File: leetcode/java/0409. Longest Palindrome.java
package leetcode.java;

import java.util.HashMap;

// Time complexity: O(N)
// Space complexity: O(N)

class Solution {
    public int longestPalindrome(String s) {
        HashMap<Character, Integer> map = new HashMap<>();

        for (char c : s.toCharArray()) {
            if (map.containsKey(c)) {
                map.put(c, map.get(c)+1);
            } else {
                map.put(c, 1);
            }
        }

        int odd = 0;
        int even = 0;

        for (char c2 : map.keySet()) {
            if (map.get(c2) % 2 == 0) {
                even += map.get(c2);
            } else {
                if (odd > 1) {
                    even += odd - 1;
                }
                odd = Integer.max(odd, map.get(c2));
            }
        }
        return odd + even;
    }
}

File: leetcode/python/0409. Longest Palindrome.py
# Time complexity: O(N)
# Space complexity: O(N)

class Solution(object):
    def longestPalindrome(self, s):

        freq = {}
        for char in s:
            if char in freq:
                freq[char] += 1
            else:
                freq[char] = 1
        
        odd = 0
        even = 0

        for al,count in freq.items():
            if count % 2 == 0:
                even += count
            else:
                if odd == 0:
                    odd = count
                elif odd > count:
                    even += count - 1
                else:
                    even += odd - 1
                    odd = count

        return odd + even